common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -0,0 +1,171 @@
+      this.comparator = comparator;
+      this.version = version;
+      this.hashFunction = hashFunction;
+      this.isBinaryCollation = isBinaryCollation;
+
+      if (isBinaryCollation) {
+        this.equalsFunction = UTF8String::equals;
+      } else {
+        this.equalsFunction = (s1, s2) -> this.comparator.compare(s1, s2) == 0;
+      }
+    }
+
+    /**
+     * Constructor with comparators that are inherited from the given collator.
+     */
+    public Collation(
+      String collationName, Collator collator, String version, boolean isBinaryCollation) {
+      this(
+        collationName,
+        collator,
+        (s1, s2) -> collator.compare(s1.toString(), s2.toString()),
+        version,
+        s -> (long)collator.getCollationKey(s.toString()).hashCode(),
+        isBinaryCollation);
+    }
+  }
+
+  private final Collation[] collatorTable;
+  private final HashMap<String, Integer> collationNameToIdMap = new HashMap<>();
+
+  private CollationFactory() {
+    collatorTable = new Collation[4];
+
+    // Binary comparison. This is the default collation.
+    // No custom comparators will be used for this collation.
+    // Instead, we rely on byte for byte comparison.
+    collatorTable[0] = new Collation(
+      "UCS_BASIC",
+      null,
+      UTF8String::compareTo,
+      "1.0",
+      s -> (long)s.hashCode(),
+      true);
+
+    // Case-insensitive UTF8 binary collation.
+    // TODO: Do in place comparisons instead of creating new strings.
+    collatorTable[1] = new Collation(
+      "UCS_BASIC_LCASE",
+      null,
+            Comparator.comparing(UTF8String::toLowerCase), "1.0",
+      (s) -> (long)s.toLowerCase().hashCode(),
+      false);
+
+    // UNICODE case sensitive comparison (ROOT locale, in ICU).
+    collatorTable[2] = new Collation(
+      "UNICODE", Collator.getInstance(ULocale.ROOT), "153.120.0.0", true);
+    collatorTable[2].collator.setStrength(Collator.TERTIARY);
+
+
+    // UNICODE case-insensitive comparison (ROOT locale, in ICU + Secondary strength).
+    collatorTable[3] = new Collation(
+            "UNICODE_CI", Collator.getInstance(ULocale.ROOT), "153.120.0.0", false);
+    collatorTable[3].collator.setStrength(Collator.SECONDARY);
+
+    for (int i = 0; i < collatorTable.length; i++) {
+      this.collationNameToIdMap.put(collatorTable[i].collationName, i);
+    }
+  }
+
+  private static final CollationFactory instance = new CollationFactory();
+
+  /**
+   * Returns the collation id for the given collation name.
+   */
+  public int collationNameToId(String collationName) throws SparkException {
+    String normalizedName = collationName.toUpperCase();
+    if (collationNameToIdMap.containsKey(normalizedName)) {
+      return collationNameToIdMap.get(normalizedName);
+    } else {
+      throw new SparkException(
+        "COLLATION_INVALID_NAME", Collections.singletonMap("collationName", collationName));
+    }
+  }
+
+  public Collation fetchCollationInfo(int collationId) {
```

Comment:
maybe refactor the names of these two methods as well if to match the new class name

Reply:
right, thanks.